### ISSUE TITLE
Fail upload if a file in the workspace or compendium is not unicode

### ISSUE THREAD

**Issue by nuest** [o2r-project/o2r-loader#10]
The extraction is extremely tricky if a file that is not unicode is given, and we also want to be able to handle ÜÄÖ etc. in names and so forth. Therefore we take the easy way out and demand that all files which we want to read in a workspace must be [Unicode](https://en.wikipedia.org/wiki/Unicode), so [one of the UFT-x encodings](https://stackoverflow.com/questions/643694/utf-8-vs-unicode).

Use https://www.npmjs.com/package/detect-character-encoding for implementation.

@7048730 Can we have a regex for all the files that the extractor actually reads? This way we could for example accept uploads that have non-unicode data files...

**Comment by ghost** [comment]
I think we're okay, if we check that for text formats,
so if you need a re something like `(.*\.(txt|rmd|r|tex|json|yml|yaml))` with ignorecase flag should do

**Comment by nuest** [comment]
Or we could check the mimetype if we have sth. text-ish.

**Comment by ghost** [comment]
mimetype check on R files might return `None`. If you can catch all text-ish files with that in js, I think it is a good solution

**Comment by LukasLohoff** [comment]
@7048730 Does the extractor read all directories (recursive) or is there some sort of maximum folder depth?

**Comment by LukasLohoff** [comment]
@nuest I think we have to find out all encodings that are forbidden / all incompatible encodings (from the list below).

The npm package mentioned above often only detects a subset of the UTF-8 encoding. For example, the `bagit.txt` (which is UTF-8) is recognized as `ISO-8859-1` - probably because it doesn't contain enough different characters to be identified as UTF-8

Encodings detected by the package:

-> see https://www.npmjs.com/package/detect-character-encoding#supported-character-sets

Edit: Seems to be only an issue for the files generated by bagit.py

**Comment by nuest** [comment]
@LukasLohoff so just for my understanding, you have created a test dataset where correct/wrong encodings are correctly detected?

If files generated by bagit.py are the issue, we can safely ignore them and only check files in the payload directory. Does that solve the problem?

**Comment by ghost** [comment]
@LukasLohoff unlimited recursiveness

**Comment by LukasLohoff** [comment]
@nuest I've tested a few different encodings, but not all. Only the bagit files were detected incorrectly (example above), but that should not be a problem if we add all unicode compatible subsets (I think only ISO-8859-1) to the list of supported encodings.

**Comment by nuest** [comment]
Ok, let's hope this list does not get too long :-). Please continue as suggested, by also accepting compatible subsets.

**Comment by nuest** [comment]
@LukasLohoff You need to take a look at this again. @MarkusKonk found that using the assemblytainer, which contains and `.rds` file, fails with the following error message:

```
{"message":"Files with unsupported encoding detected. Only UTF-8 is supported.","files":[{"file":"FKoE7/data/SpatiallyConstrainedClustering.rds","encoding":"windows-1252"}]}
```

I suggest to fix this by correcting [the regex for file detection](https://github.com/o2r-project/o2r-loader/pull/20/files#diff-ff1772f82f6193393f756ba18ae4288aR98) to use word boundaries [or something](https://stackoverflow.com/questions/18086473/regular-expression-for-extension-of-file) to make sure that `.r` as an extension matches, but `.rds` does not.

Please also update the documentation to make clear that this checking focusses on fails that are _relevant for metadata extraction_. Maybe even rename the settings to reflect this better. Because we are not analysing any imaginable text file format. Update error message accordingly (e.g. "Input file for metadata extraction with unsupported encoding detected. For these files only UTF-8 is supported.")

**Comment by LukasLohoff** [comment]
@nuest 

That's a simple fix: Adding "$" to the regex ensures the string ends after ".Rmd" / ".R".

```
passon.textFiles
Array[4] ["/tmp/o2r/compendium/DtzAj/data/Hauffe.R", "/tmp/o2r/compendium/DtzAj/data/auxiliaryFunctions.…", "/tmp/o2r/compendium/DtzAj/data/dbSTEPCAM.R" …]
length:4
__proto__:Array[0]
0:"/tmp/o2r/compendium/DtzAj/data/Hauffe.R"
1:"/tmp/o2r/compendium/DtzAj/data/auxiliaryFunctions.R"
2:"/tmp/o2r/compendium/DtzAj/data/dbSTEPCAM.R"
3:"/tmp/o2r/compendium/DtzAj/data/paper.Rmd"
``` 

Since we are supporting ISO-8859-1 because it's a Unicode subset, we could also support Windows-1252. They are basically the same: http://www.i18nqa.com/debug/table-iso8859-1-vs-windows-1252.html

**Comment by nuest** [comment]
We're supporting any character set (so that chinese people can also upload texts), but the _encoding_ must be UTF-8. Please create a PR with the fix, sounds simple enough!

**Comment by LukasLohoff** [comment]
@nuest can be closed (see #23, #30)
